### PR TITLE
[BEAM-79] add Gearpump runner

### PR DESCRIPTION
**Pull request by manuzhang**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This PR adds Gearpump runner to Beam meeting the goals of phase 1 in the [design document](https://docs.google.com/document/d/1nw64QUWVfT8L7FUprPGLEeNjSBpDMkn1otfLt2rHM5g/edit).

The Gearpump runner supports the following functionalities,

* Transformations: ParDo, GroupByKey, Flatten
* Windows: using Beam's window logic and code
* side outputs
* serialization/deserialization: using Gearpump's Kryo serializer
* sources: Beam's UnboundedSource
* message delivery guarantee: at-most-once
* tests: integration test for various translators

Here's a snapshot of running the following Beam example on Gearpump cluster

```java
    PCollection<KV<String, Long>> wordCounts =
        p.apply(Read.from(new UnboundedTextSource()).named("WordStream"))
            .apply(ParDo.of(new ExtractWordsFn()))
            .apply(Window.<String>into(FixedWindows.of(Duration.standardSeconds(10))))
            .apply(Count.<String>perElement());

    wordCounts.apply(ParDo.of(new FormatAsStringFn()));
```

snip20160511_4

Note that the Gearpump runner is still in early stage and lacking capabilities like trigger, side inputs, aggregator. However, I'd like to have the community to get a feel of what Gearpump is like, whether Beam and Gearpump go well, and gather ideas for improvements.

### CODE REVIEW

**Comment by jbonofre** [comment]
Good job. Let me take a look and try it. Thanks @manuzhang !

**Comment by davorbonaci** [comment]
Wow! We didn't expect it this fast!

We'll take a look shortly, but it might take us a few days to parse through.

R: @davorbonaci

**Comment by manuzhang** [comment]
Gearpump requires Java 8 so the test matrix for Java 7 will fail.

**Comment by davorbonaci** [comment]
Also, R: @kennknowles.

File: runners/gearpump/pom.xml [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                <exclusion>
+                    <groupId>org.apache.gearpump</groupId>
+                    <artifactId>gearpump-experimental-cgroup_2.11</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.beam</groupId>
+            <artifactId>java-sdk-all</artifactId>
+            <version>${beam.version}</version>
```
Comment (kennknowles) [inline runners/gearpump/pom.xml]: I believe you can omit this version and let the parent `<dependencyManagement>` sections set it up.

File: runners/gearpump/pom.xml [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+                <exclusion>
+                    <groupId>com.google.collections</groupId>
+                    <artifactId>google-collections</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>${junit.version}</version>
```
Comment (kennknowles) [inline runners/gearpump/pom.xml]: This version, too.

File: runners/gearpump/pom.xml [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest-all</artifactId>
+            <version>${hamcrest.version}</version>
```
Comment (kennknowles) [inline runners/gearpump/pom.xml]: And this version, too.

File: runners/pom.xml [hunk 4]
```diff
@@ -39,6 +39,7 @@
     <module>direct-java</module>
     <module>flink</module>
     <module>spark</module>
+    <module>gearpump</module>
```
Comment (kennknowles) [inline runners/pom.xml]: Since gearpump requires Java 8, you can use maven profiles to only include this module when the JDK is 8 or higher. Or perhaps it is just certain tests that require it, so you could limit the effect of the profile to exactly what needs to be disabled. We do this with the `sdks/java/java8tests` module, if you want an example.

File: runners/gearpump/pom.xml [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            </plugin>
+
+            <!-- Java compiler -->
+            <plugin>
+
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.1</version><!--$NO-MVN-MAN-VER$-->
+                <configuration>
+                    <source>${java.version}</source>
```
Comment (kennknowles) [inline runners/gearpump/pom.xml]: I believe that 1.7 is the default throughout the project via `<pluginManagement>`, is that not true?

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineResult.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+public class GearpumpPipelineResult implements PipelineResult {
+  @Override
+  public State getState() {
+    return null;
+  }
+
+  @Override
+  public <T> AggregatorValues<T> getAggregatorValues(Aggregator<?, T> aggregator)
+      throws AggregatorRetrievalException {
+    return null;
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineResult.java]: I think this should probably throw `AggregatorRetrievalException` or `UnsupportedOperationException`.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  public static GearpumpPipelineRunner fromOptions(PipelineOptions options) {
+    return new GearpumpPipelineRunner((GearpumpPipelineOptions) options);
+  }
+
+
+  public <OutputT extends POutput, InputT extends PInput> OutputT apply(
+      PTransform<InputT, OutputT> transform, InputT input) {
+    if (Window.Bound.class.equals(transform.getClass())) {
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java]: Nice correct use of `PipelineRunner#apply`. I wanted to call out that in the future Runner API proposal, we will want to migrate all of this to a phased approach where this method is removed and you do the replacement after the user has finished constructed the graph. It will be basically the same for you, just doing this sort of thing using `Pipeline.replace` in your `#run()` method.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+      return (OutputT) ((PCollection) input).apply(
+          new AssignWindowsAndSetStrategy((Window.Bound) transform));
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java]: @tgroh points out that if you use `super.apply(new AssignWindowsAndSetStrategy, input)` then the representation of the graph will still have a `Window.Bound` transform, and this will be embedded as a sub-transform implementing it. It is a bit better, as the top-level graph is close to what the user wrote.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+    private final Window.Bound<T> wrapped;
+
+    public AssignWindowsAndSetStrategy(Window.Bound<T> wrapped) {
+      this.wrapped = wrapped;
+    }
+
+    @Override
+    public PCollection<T> apply(PCollection<T> input) {
+      WindowingStrategy<?, ?> outputStrategy =
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java]: Since your `GroupByKey` implementation does not do window merging, you should reject such windowing strategies here.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/examples/StreamingWordCount.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+import org.apache.gearpump.cluster.client.ClientContext;
+import org.joda.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * streaming word count example on Gearpump runner.
+ */
+public class StreamingWordCount {
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/examples/StreamingWordCount.java]: I think we should reuse the top-level Beam examples and just set `--runner=GearpumpRunner`

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/GroupByKeyTranslator.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+/**
+ * {@link GroupByKey} is translated to Gearpump groupBy function.
+ */
+public class GroupByKeyTranslator<K, V> implements TransformTranslator<GroupByKey<K, V>> {
+  @Override
+  public void translate(GroupByKey<K, V> transform, TranslationContext context) {
+    JavaStream<WindowedValue<KV<K, V>>> inputStream =
+        context.getInputStream(context.getInput(transform));
+    int parallelism = context.getPipelineOptions().getParallelism();
+    JavaStream<WindowedValue<KV<K, Iterable<V>>>> outputStream = inputStream
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/GroupByKeyTranslator.java]: This implementation is good for non-merging windows. Just noting that you should throw an `UnsupportedOperationException` during graph construction if the windowing strategy is merging.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/functions/DoFnFunction.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        new NoOpStepContext(),
+        windowingStrategy
+    );
+  }
+
+  @Override
+  public Iterator<WindowedValue<OutputT>> apply(WindowedValue<InputT> value) {
+    outputs = Lists.newArrayList();
+
+    doFnRunner.startBundle();
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/functions/DoFnFunction.java]: Single-element bundles could be expensive.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/io/GearpumpSource.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+ * wrapper over UnboundedSource for Gearpump DataSource API.
+ */
+public class GearpumpSource<OutputT, CheckpointMarkT extends UnboundedSource.CheckpointMark>
+    implements DataSource {
+
+  private final UnboundedSource.UnboundedReader<OutputT> reader;
+
+  private boolean available = false;
+
+  public GearpumpSource(UnboundedSource<OutputT, CheckpointMarkT> source, PipelineOptions options) {
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/io/GearpumpSource.java]: You should actually extract a `Reader` at execution time, not construction time. It is probably most analogous to `open()` here, though I do not know the underlying Gearpump paradigm and these things rarely line up perfectly. CC @dhalperi

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/utils/GearpumpDoFnRunner.java [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+      Preconditions.checkNotNull(tag, "TupleTag passed to sideOutputWithTimestamp cannot be null");
+      sideOutputWindowedValue(tag, output, timestamp, null, PaneInfo.NO_FIRING);
+    }
+
+    @Override
+    protected <AggInputT, AggOutputT> Aggregator<AggInputT, AggOutputT> createAggregatorInternal(
+        String name, Combine.CombineFn<AggInputT, ?, AggOutputT> combiner) {
+      Preconditions.checkNotNull(combiner,
+          "Combiner passed to createAggregator cannot be null");
+      return null;
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/utils/GearpumpDoFnRunner.java]: Interesting case here. If aggregators are optional "extra" stuff, then you should go ahead and run the main pipeline even when they are not supported. But if they are critical to the user's success, then you should throw `UnsupportedOperationException`.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/utils/NoOpStepContext.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+  public <T, W extends BoundedWindow> void writePCollectionViewData(TupleTag<?> tag,
+      Iterable<WindowedValue<T>> data,
+      Coder<Iterable<WindowedValue<T>>> dataCoder, W window, Coder<W> windowCoder) throws
+      IOException {
+
+  }
+
+  @Override
+  public StateInternals<?> stateInternals() {
+    return null;
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/utils/NoOpStepContext.java]: Probably best to throw `UnsupportedOperationException` in cases like these. When we start exposing state to users (not just SDKs) then you'll have some static analysis of the graph to allow your runner to exclude those pipelines.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/io/GearpumpSource.java (hunk 13 above)
Comment (dhalperi) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/io/GearpumpSource.java]: specifically, it looks like this needs to happen in `open()` below.

**Comment by kennknowles** [comment]
This is really cool!

One thing that you can do to get a lot of automatic feedback on your runner is set up the `RunnableOnService` tests as in #319 (Direct), #291 (Flink), #294 (Spark). This will also help to drive which test categories we need to split into, so you can explicitly exclude tests of capabilities that are still under development.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java (hunk 7 above)
Comment (manuzhang) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java]: Sure, I basically followed the DataflowPipeline.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java (hunk 8 above)
Comment (manuzhang) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java]: oh, I originally did that but I heard test matrix complaining about type issues. Let me try again.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/examples/StreamingWordCount.java (hunk 10 above)
Comment (manuzhang) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/examples/StreamingWordCount.java]: Agreed but GearpumpRunner only supports a subset of capabilities now so not all examples will work.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/functions/DoFnFunction.java (hunk 12 above)
Comment (manuzhang) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/functions/DoFnFunction.java]: I think we process one record at a time in streaming. What's the suggested way here ?

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/io/GearpumpSource.java (hunk 13 above)
Comment (manuzhang) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/io/GearpumpSource.java]: yes, it should happen in `open()`.

**Comment by manuzhang** [comment]
@kennknowles Thanks, I'll go through all the stuff with your comments.

File: runners/pom.xml (hunk 4 above)
Comment (manuzhang) [inline runners/pom.xml]: Although Gearpump requires Java 8, I don't think either Beam application or the tests need be compiled in Java 8 to be run on Gearpump. That said, let me try this to make test matrix happy.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/io/GearpumpSource.java (hunk 13 above)
Comment (manuzhang) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/io/GearpumpSource.java]: @dhalperi @kennknowles I run into a `NotSerializableException` here. Gearpump `DataSource` is serialized into configuration and deserialized at Task side but `ProxyInvocationHandler` of `PipelineOption` is not serializable.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/functions/DoFnFunction.java (hunk 12 above)
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/functions/DoFnFunction.java]: Well, it is perfectly correct to have bundles of any size, and it is up to the runner to choose the size that works best. The bundle is the unit of atomicity. Other than the runner's particular performance decisions, users are encouraged to do potentially expensive set up in `startBundle` and flushing operations in `finishBundle`. This definitely is not a blocker.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/examples/StreamingWordCount.java (hunk 10 above)
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/examples/StreamingWordCount.java]: That is totally fine. We will track the capabilities in the matrix, and we can add commentary to the documentation about which runners work with which examples.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java (hunk 8 above)
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpPipelineRunner.java]: Any luck?

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/io/GearpumpSource.java (hunk 13 above)
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/io/GearpumpSource.java]: Serialization of `PipelineOptions` is done via Jackson. You can see any example [in the `DataflowPipelineTranslator](https://github.com/apache/incubator-beam/blob/master/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L406) and some brief documentation [in `PipelineOptions` javaodc](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java#L178).

**Comment by kennknowles** [comment]
Sorry for the slow replies!

**Comment by manuzhang** [comment]
@kennknowles updated most parts against your comments. It seems `RunnableOnService` tests use `PAssert` which requires support of `sideInput`. Unfortunately, `sideInput` is not yet supported in Gearpump runner so I have to disable the `RunnableOnService` test for now.

**Comment by manuzhang** [comment]
@kennknowles any more comments ?

**Comment by kennknowles** [comment]
Yes, sorry. I failed to click the "comment" button after writing this:

Good news! In #437 I switched `PAssert` so it no longer uses side inputs in most cases, but instead uses globally-windowed `GroupByKey`. Now almost all runners will be able to use `PAssert` from an early stage. You should be able to use `TestPipeline` and `PAssert` to test your runner using the test suite we provide.

But even before that, we can merge such a major development as this new runner into a feature branch as soon as it has some viability, so everyone has a place to collaborate. Since this runner does have tests that run some real pipelines, that is enough for a feature branch. After that, the criteria for merging into master are outlined in [this email from Frances](http://mail-archives.apache.org/mod_mbox/incubator-beam-dev/201605.mbox/%3CCAAzyFAxcmexUQnbF%3DYk0PLmM3f5e5BqwJz4%2BC5doRucLNxOh8w%40mail.gmail.com%3E).

So, first, we need to get the tests happy then we can merge into a feature branch. I will also give this another review right away.

**Comment by manuzhang** [comment]
Thanks. Did you just revert the `PAssert` changes ? And Will you create a feature branch ?

**Comment by manuzhang** [comment]
@kennknowles I assume you will support this and also count as a contributor for maintaining it :smiley_cat: ?

**Comment by kennknowles** [comment]
Yes, I had to roll back the `PAssert` changes while I work out some issues that only showed up in the postcommit integration tests. Fixing those tests turned up issues in the unit tests so I needed some time. Hopefully it will be simple and rolled forwards soon.

As for the feature branch, I'll pull this PR straight to a feature branch but I think we should get the tests green first, even though it won't block mainline development.

**Comment by kennknowles** [comment]
The `PAssert` changes are rolled forwards again, this time for sure.

**Comment by kennknowles** [comment]
Just a heads up: you are going to need to add a dependency on `runners/core-java` as we are in the process of moving utilities there.

```
<dependency>
  <groupId>org.apache.beam</groupId>
  <artifactId>beam-runners-core-java</artifactId>
</dependency>
```

**Comment by manuzhang** [comment]
@kennknowles Yes, I saw that. I'm trying out the new `PAssert` now.

**Comment by kennknowles** [comment]
Let me know if there is anything else I can make easier here. Happy to chat, etc.

**Comment by manuzhang** [comment]
@kennknowles It seems some tests are still implemented with `PCollectionViewAssert` which requires side input. Another problem is I get tests timeout on travis while I can get them passed locally.

**Comment by kennknowles** [comment]
**Tests using `PCollectionViewAssert`**: Filtering them out is the right thing to do. The way you have it is fine to get unblocked. It would be even better to tag specific test methods with `@Category(StaticSideInputs.class)` (from the [capability matrix doc](https://docs.google.com/document/d/1fICxq32t9yWn9qXhmT07xpclHeHX2VlUyVtpi2WzzGM/edit?usp=sharing) shared by @tgroh) and then use `<excludedGroups>` in your pom. This way you aren't eliminating the whole test class, just the method.

**Test timeouts**: It may be infeasible to run all these `RunnableOnService` tests in the presubmit and in Travis. For Dataflow, we run just a couple in the presubmit and then the rest as a [post-submit Jenkins job](https://builds.apache.org/job/beam_PostCommit_RunnableOnService_GoogleCloudDataflow). It probably makes sense to do this for many runners, so that is also no problem. Do you know whether it is just overheads or whether there is a real performance problem?

For right now, there's a [RAT plugin failure](https://builds.apache.org/job/beam_PreCommit_MavenVerify/2177/org.apache.beam$beam-parent/console) in the `runners-parent` module. If we fix that, then then Jenkins should be willing to run your tests to completion even if it takes a long time.

From there, since traffic on the branch will be low enough, I think we can probably check it in to a feature branch and then set up the postsubmit job and try to see if we can speed up the tests.

**Comment by kennknowles** [comment]
It looks like the RAT plugin was seeing files that needed to be cleaned up from another branch. We've just adjusted the Jenkins configuration so those files should be cleaned. Go ahead and force a rebuild via rebase and force push, or dummy commit, and we should see Jenkins succeed or timeout.

**Comment by manuzhang** [comment]
The travis build has errors as below which cannot be reproduced on my laptop.

```
java.lang.RuntimeException: Failed to construct instance from factory method TestGearpumpRunner#fromOptions(interface org.apache.beam.sdk.options.PipelineOptions)
	at org.apache.beam.sdk.util.InstanceBuilder.buildFromMethod(InstanceBuilder.java:236)
	at org.apache.beam.sdk.util.InstanceBuilder.build(InstanceBuilder.java:165)
	at org.apache.beam.sdk.runners.PipelineRunner.fromOptions(PipelineRunner.java:59)
	at org.apache.beam.sdk.testing.TestPipeline.fromOptions(TestPipeline.java:103)
	at org.apache.beam.sdk.testing.TestPipeline.create(TestPipeline.java:98)
	at org.apache.beam.sdk.transforms.CreateTest.testCreateTimestampedEmpty(CreateTest.java:261)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:168)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runners.Suite.runChild(Suite.java:127)
	at org.junit.runners.Suite.runChild(Suite.java:26)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:161)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor30.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.apache.beam.sdk.util.InstanceBuilder.buildFromMethod(InstanceBuilder.java:225)
	... 41 more
Caused by: java.util.concurrent.TimeoutException: Futures timed out after [10000 milliseconds]
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
	at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:190)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
	at scala.concurrent.Await$.result(package.scala:190)
	at akka.remote.Remoting.start(Remoting.scala:181)
	at akka.remote.RemoteActorRefProvider.init(RemoteActorRefProvider.scala:190)
	at akka.cluster.ClusterActorRefProvider.init(ClusterActorRefProvider.scala:31)
	at akka.actor.ActorSystemImpl.liftedTree2$1(ActorSystem.scala:666)
	at akka.actor.ActorSystemImpl._start$lzycompute(ActorSystem.scala:663)
	at akka.actor.ActorSystemImpl._start(ActorSystem.scala:663)
	at akka.actor.ActorSystemImpl.start(ActorSystem.scala:679)
	at akka.actor.ActorSystem$.apply(ActorSystem.scala:141)
	at akka.actor.ActorSystem$.apply(ActorSystem.scala:118)
	at org.apache.gearpump.cluster.embedded.EmbeddedCluster.start(EmbeddedCluster.scala:51)
	at org.apache.beam.runners.gearpump.TestGearpumpRunner.<init>(TestGearpumpRunner.java:40)
	at org.apache.beam.runners.gearpump.TestGearpumpRunner.fromOptions(TestGearpumpRunner.java:48)
	... 45 more
```

**Comment by kennknowles** [comment]
Curious. I can't say much about the details there. What was at the other end of the future that timed out when the embedded Gearpump cluster was starting up?

I'll pull and try it to see if my environment reproduces the issue.

**Comment by manuzhang** [comment]
@kennknowles finally, travis is green.
I tried `mvn clean apache-rat:check` locally and the output is 

```
1 Unknown Licenses

*******************************

Unapproved licenses:

  .github/PULL_REQUEST_TEMPLATE.md

*******************************
```

**Comment by kennknowles** [comment]
I can confirm that failure. It should certainly succeed. I'm looking into it.

**Comment by kennknowles** [comment]
From a quick search it looks like the failure of `apache-rat:check` is working as intended (for some definition of "intended") according to [this thread](https://mail-archives.apache.org/mod_mbox/creadur-dev/201305.mbox/%3C518C03C0.3070806%40blueyonder.co.uk%3E) which I found through [this issue](https://github.com/dkpro/jweb1t/issues/3).

The fix suggested is to invoke the plugin only through `mvn verify`. But if that succeeded, we would expect Jenkins to succeed, since it is currently green outside this PR.

**Comment by kennknowles** [comment]
I have identified two issues that may prevent you from finding the real problem:

1. We have only _configured_ the RAT plugin in the release profile, which Jenkins uses. We probably meant to only _execute_ it in the `release` profile, but configuration should be global. So you can get the proper excludes on the command line with `-Prelease`. I will fix this separately.
2. RAT includes defaults for IDEs to exclude things like Eclipse's `.settings` and `.classpath` but the exclude patterns are not correct in the presence of subprojects. I will also fix this.

When I fix these two issues locally, the remaining RAT failure is `runners/gearpump/README.md`. Obviously we want it. The issue is that your `README.md` needs to have the Apache license header in a comment. See the source code for any of the other `README.md` files.

I would fix this for you and merge to a feature branch, but since it is a licensing issue I think it best that you make the edits.

**Comment by manuzhang** [comment]
@kennknowles Thanks for the thorough explanation. I think we are good to go this time.

**Comment by dhalperi** [comment]
Boom! :shipit:

**Comment by kennknowles** [comment]
Awesome!

**Comment by kennknowles** [comment]
Merged to branch `gearpump-runner`. Since the PR is against `master` that didn't close it, but feel free to close any time.

**Comment by manuzhang** [comment]
😄
